extensions/amp-story-shopping/0.1/test/test-amp-story-shopping-attachment.js

```diff
@@ -26,113 +28,200 @@ describes.realWin(
+        .stub(attachmentChildImpl, 'mutateElement')
+        .callsFake((fn) => Promise.resolve(fn()));
 
-      shoppingImpl = await shoppingEl.getImpl();
+      // Set page to active.
+      storeService.dispatch(Action.CHANGE_PAGE, {id: 'page1', index: 1});
     });
 
-    async function dispatchTestShoppingData() {
-      const shoppingData = {
-        'lamp': {
-          'productId': 'lamp',
-          'productTitle': 'Brass Lamp',
-          'productBrand': 'Lamp Co',
-          'productPrice': 799.0,
-          'productPriceCurrency': 'USD',
-          'productImages': ['https://source.unsplash.com/Ry9WBo3qmoc/500x500'],
-        },
-      };
-      storeService.dispatch(Action.CHANGE_PAGE, {
-        id: 'page1',
-        index: 1,
+    function dispatchActiveProductData() {
+      storeService.dispatch(Action.ADD_SHOPPING_DATA, {
+        activeProductData: shoppingData.items[0],
       });
-      storeService.dispatch(Action.ADD_SHOPPING_DATA, shoppingData);
-      storeService.dispatch(Action.TOGGLE_PAGE_ATTACHMENT_STATE, true);
     }
 
     it('should build shopping attachment component', () => {
       expect(() => shoppingImpl.layoutCallback()).to.not.throw();
     });
 
-    it('should build CTA with i18n shopping label text', async () => {
-      await dispatchTestShoppingData();
-      const attachmentChildEl = shoppingEl.querySelector(
-        'amp-story-page-attachment'
-      );
+    it('should build CTA with i18n shopping label text', () => {
       expect(attachmentChildEl.getAttribute('cta-text')).to.equal('Shop Now');
     });
 
-    it('should open attachment', async () => {
-      await dispatchTestShoppingData();
-      const attachmentChildEl = shoppingEl.querySelector(
-        'amp-story-page-attachment'
-      );
-      const attachmentChildImpl = await attachmentChildEl.getImpl();
-      env.sandbox.stub(attachmentChildImpl, 'mutateElement').callsFake(() => {
-        expect(pageEl.querySelector('.i-amphtml-story-draggable-drawer-open'))
-          .to.not.be.null;
-      });
+    it('should build PLP on attachment state open if no active product data', async () => {
+      await shoppingImpl.layoutCallback();
+      await attachmentChildImpl.layoutCallback();
+      storeService.dispatch(Action.TOGGLE_PAGE_ATTACHMENT_STATE, true);
+      expect(
+        attachmentChildEl.querySelector('.i-amphtml-amp-story-shopping-plp')
+      ).to.exist;
     });
 
-    it('should build PLP on CTA click', async () => {
-      await dispatchTestShoppingData();
-      const attachmentChildEl = shoppingEl.querySelector(
-        'amp-story-page-attachment'
-      );
-      const attachmentChildImpl = await attachmentChildEl.getImpl();
-      env.sandbox.stub(attachmentChildImpl, 'mutateElement').callsFake(() => {
-        expect(pageEl.querySelector('.amp-story-shopping-plp')).to.not.be.null;
-      });
+    it('should open attachment when active product data is set', async () => {
+      await shoppingImpl.layoutCallback();
```

Comment:
More of a preference than a suggestion, but you can pull
```
 await shoppingImpl.layoutCallback();
 await attachmentChildImpl.layoutCallback();
```
 into an async helper function and just call that instead of the two lines above. That way, you don't have to repeat this code 5x.

Reply:
Great suggestion aligning with DRY principles. Thanks Jonathan!